src/devices/machine/s2350.cpp

```diff
@@ -0,0 +1,411 @@
+
+
+s2350_device::s2350_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock) :
+	device_t(mconfig, S2350, tag, owner, clock),
+	m_tbmt(*this),
+	m_fct(*this),
+	m_rda(*this),
+	m_scr(*this),
+	m_ror(*this),
+	m_rpe(*this),
+	m_tx_cb(*this)
+{
+}
+
+void s2350_device::device_start()
+{
+	save_item(NAME(m_transmit_buffer_empty));
+	save_item(NAME(m_fill_char_transmitted));
+	save_item(NAME(m_transmitter_holding_reg));
+	save_item(NAME(m_transmitter_fill_reg));
+	save_item(NAME(m_transmitter_shift_reg));
+	save_item(NAME(m_serial_tx_state));
+
+	save_item(NAME(m_received_data_available));
+	save_item(NAME(m_receiver_overrun));
+	save_item(NAME(m_receiver_parity_error));
+	save_item(NAME(m_sync_character_received));
+
+	save_item(NAME(m_receiver_output_reg));
+	save_item(NAME(m_receiver_sync_reg));
+	save_item(NAME(m_receiver_shift_reg));
+	save_item(NAME(m_serial_rx_state));
+
+	save_item(NAME(m_in_sync));
+}
+
+void s2350_device::device_reset()
+{
+	// transmitter flags
+	set_transmit_buffer_empty(true);
+	set_fill_char_transmitted(false);
+
+	// transmitter registers
+	m_transmitter_shift_reg = 0xff;
+	m_transmitter_holding_reg = 0xff;
+	m_transmitter_fill_reg = 0x00;
+
+	m_serial_tx_state = 0;
+
+	set_received_data_available(false);
+	set_receiver_overrun(false);
+	set_receiver_parity_error(false);
+	set_sync_character_received(false);
+
+	m_sync_search_active = false;
+	m_in_sync = false;
+
+	m_serial_rx_state = 0;
+
+	// receiver registers
+	m_receiver_output_reg = 0x00;
+	m_receiver_sync_reg = 0x00;
+	m_receiver_shift_reg = 0x00;
+}
+
+void s2350_device::transmitter_holding_reg_w(u8 data)
+{
+	LOGREG("%s: 0x%02x\n", FUNCNAME, data);
+
+	m_transmitter_holding_reg = data;
+	set_transmit_buffer_empty(false);
+}
+
+void s2350_device::transmit_fill_reg_w(u8 data)
+{
+	LOGREG("%s: 0x%02x\n", FUNCNAME, data);
+
+	m_transmitter_fill_reg = data;
+}
+
+void s2350_device::receiver_sync_reg_w(u8 data)
+{
+	LOGREG("%s: 0x%02x\n", FUNCNAME, data);
+
+	m_receiver_sync_reg = data;
+}
+
+u8 s2350_device::receiver_sync_search()
+{
+	set_sync_character_received(false);
+	m_sync_search_active      = true;
+
+	LOGREG("%s: 0x%02x\n", FUNCNAME, m_receiver_sync_reg);
+
+	return m_receiver_sync_reg;
+}
+
+u8 s2350_device::receiver_output_reg_r()
+{
+	set_received_data_available(false);
+
+	LOGREG("%s: 0x%02x\n", FUNCNAME, m_receiver_output_reg);
+
+	return m_receiver_output_reg;
+}
+
+u8 s2350_device::status_word_r()
+{
+	u8 status = 0x00;
+
+	status |= m_received_data_available ? 0x01 : 0x00;
+	status |= m_receiver_overrun        ? 0x02 : 0x00;
+	status |= m_receiver_parity_error   ? 0x04 : 0x00;
+	status |= m_sync_character_received ? 0x08 : 0x00;
+	status |= m_fill_char_transmitted   ? 0x40 : 0x00;
+	status |= m_transmit_buffer_empty   ? 0x80 : 0x00;
+
+	set_fill_char_transmitted(false);
+	set_receiver_overrun(false);
+	set_receiver_parity_error(false);
+	set_sync_character_received(false);
```

Comment:
Remember to check for disabled side effects before doing things with side effects in read handlers.  Failing to do this makes systems undebuggable.